usrtools/txgen/app/cmds.c

```diff
@@ -977,3 +983,168 @@ enable_capture(port_info_t *info, uint32_t state)
 {
     txgen_set_capture(info, state);
 }
+
+/**
+ *
+ * start_stop_latency_sampler - Starts or stops latency sampler.
+ *
+ * DESCRIPTION
+ * Starts or stops latency sampler.
+ *
+ * RETURNS: N/A
+ *
+ * SEE ALSO:
+ */
+
+void
+start_stop_latency_sampler(port_info_t *info, uint32_t state)
+{
+    if (state == ENABLE_STATE)
+        txgen_start_latency_sampler(info);
+    else if (state == DISABLE_STATE)
+        txgen_stop_latency_sampler(info);
+}
+
+/**
+ *
+ * txgen_start_latency_sampler - Starts latency sampler.
+ *
+ * DESCRIPTION
+ * Starts latency sampler.
+ *
+ * RETURNS: N/A
+ *
+ * SEE ALSO:
+ */
+
+void
+txgen_start_latency_sampler(port_info_t *info)
+{
+
+    /* Start sampler */
+    if (txgen_tst_port_flags(info, SAMPLING_LATENCIES))
+        return;
+
+    if (info->latsamp_rate == 0 || info->latsamp_type == LATSAMPLER_UNSPEC ||
+        info->latsamp_num_samples == 0) {
+        CNE_ERR("Set proper sampling type, number, rate and outfile!");
+        return;
+    }
+
+    info->latsamp_stats.pkt_counter = 0;
+    info->latsamp_stats.next        = 0;
+    info->latsamp_stats.idx         = 0;
+    info->latsamp_stats.num_samples = info->latsamp_num_samples;
+
+    if (info->pkt.pktSize < (CNE_ETHER_MIN_LEN - ETHER_CRC_LEN) + sizeof(tstamp_t))
+        info->pkt.pktSize += sizeof(tstamp_t);
+
+    info->pkt.ipProto = IPPROTO_UDP;
+    txgen_packet_ctor(info);
+
+    /* Start sampling */
+    txgen_set_port_flags(info, SAMPLING_LATENCIES);
+}
+
+/**
+ *
+ * txgen_stop_latency_sampler - Stops latency sampler
+ *
+ * DESCRIPTION
+ * Stops latency sampler
+ *
+ * RETURNS: N/A
+ *
+ * SEE ALSO:
+ */
+void
+txgen_stop_latency_sampler(port_info_t *info)
+{
+    FILE *outfile;
+    uint32_t i, count;
+
+    if (txgen_tst_port_flags(info, SAMPLING_LATENCIES) == 0)
+        return;
+
+    /* Stop sampling */
+    txgen_clr_port_flags(info, SAMPLING_LATENCIES);
+
+    /* Dump stats to file */
+    outfile = fopen(info->latsamp_outfile, "w");
+    if (outfile != NULL) {
+        fprintf(outfile, "Latency\n");
+        for (i = 0; i < info->latsamp_stats.idx; i++) {
+            fprintf(outfile, "%" PRIu64 "\n", info->latsamp_stats.data[i]);
+            count++;
+        }
+        fclose(outfile);
+    }
+
+    /* Reset stats data */
+    info->latsamp_stats.pkt_counter = 0;
+    info->latsamp_stats.next        = 0;
+    info->latsamp_stats.idx         = 0;
+    info->latsamp_stats.num_samples = 0;
+
+    if (info->pkt.pktSize >= (CNE_ETHER_MIN_LEN - ETHER_CRC_LEN) + sizeof(tstamp_t))
+        info->pkt.pktSize -= sizeof(tstamp_t);
+
+    info->pkt.ipProto = IPPROTO_UDP;
```

Comment:
When stopping latency why do we set it back to IPPROTO_UDP, is it something else? If we are forcing UDP and TCP is not allowed it seems strange to me. I did not write the latency code at least I do not remember, but I should have flagged this in the original code. :-(

Reply:
No worries - makes sense. I will clean it up